denops/ddc/ddc.ts

```diff
@@ -125,82 +169,105 @@ export class Ddc {
+      options.sources
+        .map((n) => this.sources[n])
+        .filter((v) => v)
+        .map(async (s) => {
+          const [sourceOptions, sourceParams] = sourceArgs(options, s);
+          const [completePos, candidates] = await Promise.all([
+            s.getCompletePosition(
+              denops,
+              context,
+              sourceOptions,
+              sourceParams,
+            ),
+            s.gatherCandidates(
+              denops,
+              context,
+              sourceOptions,
+              sourceParams,
+            ),
+          ]);
+          return {
+            completePos,
+            candidates: candidates
+              .pipeThrough(this.filterCandidates(
+                denops,
+                context,
+                sourceOptions,
+                options.filterOptions,
+                options.filterParams,
+              ))
+              .pipeThrough(this.finalizeCandidates(
+                s,
+                sourceOptions,
+              )),
+          };
+        }),
+    );
+    const completePos = Math.max(
+      ...sources.map(({ completePos }) => completePos),
+    );
+    const candidates = concatStreams<DdcCandidate[]>(
+      ...sources.map(({ candidates }) => candidates),
+    );
+    return { completePos, candidates };
   }
 
-  private async filterCandidates(
+  private filterCandidates(
     denops: Denops,
     context: Context,
     sourceOptions: SourceOptions,
     filterOptions: Record<string, Partial<FilterOptions>>,
     filterParams: Record<string, Partial<Record<string, unknown>>>,
-    cdd: Candidate[],
-  ): Promise<Candidate[]> {
+  ): TransformStream<Candidate[]> {
     const foundFilters = (names: string[]) =>
       names.map((name) => this.filters[name]).filter((x) => x);
     const matchers = foundFilters(sourceOptions.matchers);
     const sorters = foundFilters(sourceOptions.sorters);
     const converters = foundFilters(sourceOptions.converters);
 
-    for (const matcher of matchers) {
-      const [o, p] = filterArgs(filterOptions, filterParams, matcher);
-      cdd = await matcher.filter(denops, context, o, p, cdd);
-    }
-    for (const sorter of sorters) {
-      const [o, p] = filterArgs(filterOptions, filterParams, sorter);
-      cdd = await sorter.filter(denops, context, o, p, cdd);
-    }
-
-    // Filter by maxCandidates
-    cdd = cdd.slice(0, sourceOptions.maxCandidates);
+    return new TransformStream({
+      async transform(chunk, controller) {
+        for (const matcher of matchers) {
+          const [o, p] = filterArgs(filterOptions, filterParams, matcher);
+          chunk = await matcher.filter(denops, context, o, p, chunk);
+        }
+        for (const sorter of sorters) {
+          const [o, p] = filterArgs(filterOptions, filterParams, sorter);
+          chunk = await sorter.filter(denops, context, o, p, chunk);
+        }
+        // Filter by maxCandidates
+        chunk = chunk.slice(0, sourceOptions.maxCandidates);
```

Comment:
I don't think this is the right place to apply candidate threshold but I'm not sure why you've applied to so I just leave it as-is.

Reply:
> I don't think this is the right place to apply candidate threshold but I'm not sure why you've applied to so I just leave it as-is.

Why do you think it is not the right place?
Please describe it.

Reply:
Because it applies the threshold to the chunk in this version. The previous version applies to the final candidates.

Reply:
Ah, OK.  So it is not the right place.
I think `completePos` and `candidates` should not be merged in the chunk.